x/capability/keeper/keeper.go

```diff
@@ -106,25 +106,10 @@ func (k *Keeper) InitializeAndSeal(ctx sdk.Context) {
 	defer iterator.Close()
 	for ; iterator.Valid(); iterator.Next() {
 		index := types.IndexFromKey(iterator.Key())
-		cap := types.NewCapability(index)
 
 		var capOwners types.CapabilityOwners
 		k.cdc.MustUnmarshalBinaryBare(iterator.Value(), &capOwners)
-
-		for _, owner := range capOwners.Owners {
-			// Set the forward mapping between the module and capability tuple and the
-			// capability name in the memKVStore
-			memStore.Set(types.FwdCapabilityKey(owner.Module, cap), []byte(owner.Name))
-
-			// Set the reverse mapping between the module and capability name and the
-			// index in the in-memory store. Since marshalling and unmarshalling into a store
-			// will change memory address of capability, we simply store index as value here
-			// and retrieve the in-memory pointer to the capability from our map
-			memStore.Set(types.RevCapabilityKey(owner.Module, owner.Name), sdk.Uint64ToBigEndian(index))
-
-			// Set the mapping from index from index to in-memory capability in the go map
-			k.capMap[index] = cap
-		}
+		k.InitializeCapability(ctx, index, capOwners)
```

Comment:
I think we're calling this twice, on app initialization and on init genesis.

Reply:
Yea... unfortunately don't have a great way to get over this. It needs to happen in app initialization so that restarts from db works. But it also needs to happen in Genesis so we can load capabilities from genesis file into an empty db.

Shouldn't be a problem, since if it is just an app restart, then initChain doesn't run.

If its starting a new chain the `InitializeCapability` on app creation shouldn't do anything

Reply:
From simapp/app.go

> // This must be done during creation of baseapp rather than in InitChain so
// that in-memory capabilities get regenerated on app restart

Is this the case anymore?

Reply:
Yes, per @AdityaSripal's comment:

```
It needs to happen in app initialization so that restarts from db works.
```

Reply:
We shouldn't be calling `InitializeCapability` from `InitGenesis`, at least not directly.

Reply:
This is non-determinism, we don't export state so it should have nothing to do with `InitGenesis` @fedekunze. There must be an issue elsewhere.

Reply:
This looks like dev-error:

1st run:
```
Selected randomly generated capability parameters:
{
  "index": "552",
  "owners": null
}
```

2nd run:
```
Selected randomly generated capability parameters:
{
  "index": "822",
  "owners": null
}
```